README.md
(some) LaTeX environments for Jupyter notebook
==============================================

[![Join the chat at https://gitter.im/jfbercher/jupyter_latex_envs](https://badges.gitter.im/jfbercher/jupyter_latex_envs.svg)](https://gitter.im/jfbercher/jupyter_latex_envs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

This extension for Jupyter notebook enables to use some LaTeX commands and environments markdown cells. 

1. **LaTeX commands and environments**
    - support for some LaTeX commands within markdown cells, *e.g.* `\textit`, `\textbf`, `\underline`
    - support for **theorems-like environments**, support for labels and cross references
    - support for **lists**: *enumerate, itemize*,  
    - limited support for a **figure environment**,
    - support for an environment *listing*,
    - additional *textboxa* environment
2. **Citations and bibliography**
    - support for `\cite` with creation of a References section
3. **Document-wide numbering of equations and environments, support for `\label` and `\ref`**
4. **Configuration toolbar**
5. **LaTeX_envs dropdown menu for a quick insertion of environments**
6. **User's LaTeX definitions** file can be loaded and used 
7. **Export to HTML and LaTeX with a customized exporter**
8. Styles can be customized in the `latex_env.css` stylesheet

More environments can be simply added in the source file (`thmsInNb4.js`). 

It is possible to export the notebooks to plain $\LaTeX$ and html while keeping all the features of the `latex_envs` notebook extension in the converted version. We provide specialized exporters, pre and post processors, templates. We also added entry-points to simplify the conversion process. It is now as simple asIt is now as simple as
```bash
jupyter nbconvert --to html_with_lenvs FILE.ipynb
```
or 
```bash
jupyter nbconvert --to latex_with_lenvs FILE.ipynb
```
to convert `FILE.ipynb` into html/latex while keeping all the features of the `latex_envs` notebook extension in the converted version. The LaTeX converter also expose several conversion options (read the [docs](https://rawgit.com/jfbercher/jupyter_latex_envs/master/src/latex_envs/static/doc/latex_envs_doc.html). 


Demo/documentation
==================

The `doc` subdirectory that constains an example notebook and its html and pdf versions. This serves as the documentation. 
A demo notebook `latex_env_doc.ipynb` is provided. Its html version is [latex_env_doc.html](https://rawgit.com/jfbercher/jupyter_latex_envs/master/src/latex_envs/static/doc/latex_env_doc.html) and a pdf resulting 
from conversion to LaTeX is available as [documentation](https://rawgit.com/jfbercher/jupyter_latex_envs/master/src/latex_envs/static/doc/documentation.pdf). 


Installation
============

The extension consists in a package that includes a javascript notebook extension. Since Jupyter 4.2, this is the recommended way to distribute nbextensions. The extension can be installed

- from the master version on the github repo (this will be always the most recent version)
- via pip for the version hosted on Pypi
- as part of the great [Jupyter-notebook-extensions](https://github.com/ipython-contrib/Jupyter-notebook-extensions) collection. Follow the instructions there for installing. Once this is done, you can open a tab at  `http://localhost:8888/nbextensions` to enable and configure the various extensions. 

From the github repo or from Pypi, 

- **step 1**: install the package
     - `pip3 install https://github.com/jfbercher/jupyter_latex_envs/archive/master.zip [--user][--upgrade]`
     -  <span style="color:blue"> or</span> `pip3 install jupyter_latex_envs [--user][--upgrade]`
     - <span style="color:blue"> or</span> clone the repo and install
   ```
   git clone https://github.com/jfbercher/jupyter_latex_envs.git
   python3 setup.py install
   ```
 
With Jupyter >= 4.2, step 1 should sufficient as the files will be automatically copied to the target drectories and the extension enabled. For ealier versions of Jupyter or if something fails, you will have to add two more steps. 

- **step 2**: install the notebook extension 
```
jupyter nbextension install --py latex_envs [--user]
```

-  **step 3**: and enable it
```
jupyter nbextension enable latex_envs [--user] --py
```

For Jupyter versions before 4.2, the situation is more tricky since you will have to find the location of the source files (instructions from @jcb91 found [here](https://github.com/jcb91/jupyter_highlight_selected_word)): execute 
```
python -c "import os.path as p; from jupyter_highlight_selected_word import __file__ as f, _jupyter_nbextension_paths as n; print(p.normpath(p.join(p.dirname(f), n()[0]['src'])))"
```
Then, issue 
```
jupyter nbextension install <output source directory>
jupyter nbextension enable latex_envs/latex_envs
```
where `<output source directory>` is the output of the python command.  


Disclaimer, sources and acknowledgments
=======================================


Originally, I used a piece of code from the nice online markdown editor `stackedit` [https://github.com/benweet/stackedit/issues/187](https://github.com/benweet/stackedit/issues/187), where the authors also considered the problem of incorporating LaTeX markup in their markdown. 

I also studied and used examples and code from [https://github.com/ipython-contrib/IPython-notebook-extensions](https://github.com/ipython-contrib/IPython-notebook-extensions).

- This is done in the hope it can be useful. However there are many impovements possible, in the code and in the documentation. 
**Contributions will be welcome and deeply appreciated.** 

- If you have issues, please post an issue at 
`https://github.com/jfbercher/jupyter_latex_envs/issues` [here](https://github.com/jfbercher/jupyter_latex_envs/issues). 

**Self-Promotion --** Like `latex_envs`? Please star and follow the [repository](https://github.com/jfbercher/jupyter_latex_envs) on GitHub. 


history.txt

2254f79 - Jean-François Bercher, Sun Sep 18 19:37:40 2016 +0200 : Updated templates to insert user's latex_defs.tex when exporting to html/latex
1a974f6 - Jean-François Bercher, Sun Sep 18 19:35:23 2016 +0200 : When inserting an environment, selected text is inserted within the new environment
8467213 - Jean-François Bercher, Sun Sep 18 19:32:50 2016 +0200 : Added style for enumerate and for optional parameters in math environments
17f425c - Jean-François Bercher, Sun Sep 18 19:26:48 2016 +0200 : - Added support for optional parameters in environments, eg \\begin{theorem}[Option] - a class latex_title_opt is available to customize styling - Modified processing for enumerate and itemize environment in order to provide classes ul.item/li.item and ol.enum/li.enum for style customization.
4e8bde9 - Jean-François Bercher, Sun Sep 18 19:21:32 2016 +0200 : - Added loading of user's personnal LaTeX definitions - Added a menu option to load these definitions - Changed the way the system configuration/document configuration are loaded and merged (using jquery extend)
7864b2f - Jean-François Bercher, Sun Sep 18 19:19:42 2016 +0200 : Enable to include user's LaTeX definitions in html export
186778c - Jean-François Bercher, Sun Sep 18 19:14:52 2016 +0200 : Put system config (in notebook.json) into a single object latex_env
3e5ac01 - Jean-François Bercher, Sat Sep 17 17:43:26 2016 +0200 : - Enable loading of user personal LaTeX definitions - updated toogle menu to toggle loading of user's LaTeX definitions - Changed style for nested enumerate environments
1ab12f5 - Jean-François Bercher, Tue Sep 13 00:58:00 2016 +0200 : 1.2.14
503048b - Jean-François Bercher, Tue Sep 13 00:56:47 2016 +0200 : revert styling for nested enumerates
c33be80 - Jean-François Bercher, Sun Sep 11 23:45:16 2016 +0200 : version 1.2.13
2094da7 - Jean-François Bercher, Sun Sep 11 23:43:39 2016 +0200 : Added a Toogle menu in the config toolbat to: - toggle the display of the LaTeX_envs dropdown menu, - toggle the display of labels keys, - more to come...
46de6e0 - Jean-François Bercher, Sun Sep 11 23:41:10 2016 +0200 : Cross references now use the true environment number instead of the reference//label key. References are updated immediately. This works document wide and works for pre and post references
0e994a1 - jf.bercher, Thu Sep 8 14:59:49 2016 +0200 : add support for LaTeX of optional parameters in theorems (but not in html for now)
5450973 - jf.bercher, Thu Sep 8 14:58:21 2016 +0200 : added support for python 2 in setup
9be1ccc - jf.bercher, Wed Sep 7 19:03:04 2016 +0200 : minor cosmetic changes
501fecf - jf.bercher, Wed Sep 7 18:52:56 2016 +0200 : minor cosmetic changes
476d392 - jf.bercher, Wed Sep 7 13:56:13 2016 +0200 : modified install procedure to 'one shot'
467be29 - jf.bercher, Mon Sep 5 11:50:26 2016 +0200 : 1.2.9 modified setup.py to show an installation message
a89c95b - jf.bercher, Mon Sep 5 10:40:21 2016 +0200 : 1.2.8 removed reference to pypandoc
a916987 - Jean-François Bercher, Sun Sep 4 23:54:06 2016 +0200 : version 1.2.7
e50da3e - Jean-François Bercher, Sun Sep 4 23:53:05 2016 +0200 : Added css for nested numbered lists
25efdfb - Jean-François Bercher, Sun Sep 4 23:52:12 2016 +0200 : issue #3: removed support for \\, added \par, corrected <,
7ea9842 - Jean-François Bercher, Sat Sep 3 10:16:05 2016 +0200 : version 1.2.6
00e9401 - Jean-François Bercher, Sat Sep 3 10:11:09 2016 +0200 : updated documentation
4009776 - Jean-François Bercher, Sat Sep 3 10:10:37 2016 +0200 : updated .gitignore
8491c56 - Jean-François Bercher, Sat Sep 3 09:59:21 2016 +0200 : corrected linebreaks issue #1 + support for % and \\
efbdfd5 - Jean-François Bercher, Sat Sep 3 09:58:41 2016 +0200 : corrected nebreaks issue #1
a9a4682 - Jean-François Bercher, Sat Sep 3 09:57:51 2016 +0200 : improved conversion to latex
e8d1f7e - Jean-François Bercher, Fri Sep 2 00:49:38 2016 +0200 : corrected $$  in environments interpreted as $
c6e9f7d - Jean-François Bercher, Fri Sep 2 00:45:48 2016 +0200 : version 1.2.5
5b28935 - Jean-François Bercher, Fri Sep 2 00:44:05 2016 +0200 : corrected 22416  in environments interpreted as $

src/latex_envs/static/initNb.js

// Initializations

function onMarkdownCellRendering(event, data) {
    // console.log("recomputing eqs")
    if (MathJaxDefined) MathJax.Hub.Queue(
        ["resetEquationNumbers", MathJax.InputJax.TeX], 
        ["PreProcess", MathJax.Hub], 
        ["Reprocess", MathJax.Hub]
    );
    $('.latex_label_anchor').toggle(labels_anchors); 
};

function toggleLatexMenu() {
    if (!LaTeX_envs_menu_present) {
        $('#Latex_envs').remove();
        $('#toggleLatexMenu').css('color', 'grey')
            .attr('title', 'Insert LaTeX_envs menu');
    } else {
        create_latex_menu();
        $('#toggleLatexMenu').css('color', 'black')
            .attr('title', 'Remove LaTeX_envs menu');
    }
}

function loadLatexUserDefs() {
    $.get('latexdefs.tex').done(function(data) {
        data = data.replace(/^/gm, '\$\$\$').replace(/$/gm, '\$\$\$');
        if ($('#latexdefs').length > 0) $('#latexdefs').remove();
        $('body').append($('<div/>').attr('id', 'latexdefs').text(data));
        console.log('latex_envs: loaded user LaTeX definitions latexdefs.tex');
        onMarkdownCellRendering();
    }).fail(function() {
        console.log('latex_envs: failed to load user LaTeX definitions latexdefs.tex')
    });
    
}

var init_nb = function() {
    readBibliography(function() {
        create_latex_menu();
        if (!LaTeX_envs_menu_present) toggleLatexMenu();
        if (latex_user_defs) loadLatexUserDefs();
        add_help_menu_item();
        createReferenceSection();
        init_cells();
        Jupyter.keyboard_manager.edit_shortcuts.add_shortcuts(add_edit_shortcuts);
    });
}


var init_cells = function() {
    var ncells = Jupyter.notebook.ncells();
    var cells = Jupyter.notebook.get_cells();
    var MarkdownCell = require('notebook/js/textcell').MarkdownCell;
    var maps = initmap(); // this is to reset the counters in case of reload
    var venvironmentMap = maps[0];
    var vcit_table = maps[3];
    environmentMap = maps[0];
    cit_table = maps[3];
    cmdsMap = maps[1];
    eqLabNums = maps[2];
    eqNum = eqNumInitial;
    current_cit = current_citInitial;
    var noevent = true;
    var lastmd_cell;
    for (var i = 0; i < ncells; i++) {
        var cell = cells[i]; 
        if (cell instanceof MarkdownCell) {
            cell.render(noevent);
            lastmd_cell = cell;
        };
    }
    if(typeof lastmd_cell !== "undefined") lastmd_cell.render(); // re-render last md cell and issue rendered.MarkdownCell event
    onMarkdownCellRendering();
    //$('.latex_label_anchor').toggle(labels_anchors); 
}


// ** load configuration (1) default config, (2) system config (3) document config
// and save at document level
function init_config(Jupyter,utils,configmod) {
    var cfg = { // default config
            //EQUATIONS
            'eqNumInitial': 1, // begins equation numbering at eqNum
            'eqLabelWithNumbers': true, //if true, label equations with equation numbers; otherwise using the tag specified by \label
            //BIBLIOGRAPHY
            'current_citInitial': 1, // begins citation numbering at current_cit
            'cite_by': 'apalike', //cite by 'number', 'key' or 'apalike' 
            'bibliofile': 'biblio.bib', //or IPython.notebook.notebook_name.split(".")[0]+."bib"
            // LaTeX_envs_menu 
            'LaTeX_envs_menu_present': true,
            // Show anchors for labels
            'labels_anchors':false,
            // Load LaTeX user definitions
            'latex_user_defs': false
        }
    // create config object to load parameters
    var base_url = utils.get_body_data("baseUrl");
    var config = new configmod.ConfigSection('notebook', { base_url: base_url });
    config.loaded.then(function config_loaded_callback() {
        // config may be specified at system level or at document level.
        // first, update defaults with config loaded from server
        cfg = $.extend(true, cfg, config.data.latex_envs);
        // then update cfg with any found in current notebook metadata
        // and save in nb metadata (then can be modified per document)
        cfg = IPython.notebook.metadata.latex_envs = $.extend(true, cfg,
            IPython.notebook.metadata.latex_envs);
        // update global variables
        cite_by = cfg.cite_by; //global
        bibliofile = cfg.bibliofile;
        eqNumInitial = cfg.eqNumInitial;
        eqLabelWithNumbers = cfg.eqLabelWithNumbers;
        eqNum = cfg.eqNumInitial;
        LaTeX_envs_menu_present = cfg.LaTeX_envs_menu_present;
        labels_anchors = cfg.labels_anchors; 
        latex_user_defs = cfg.latex_user_defs;
        reprocessEqs = true;
        // and initialize bibliography and cells 
        init_nb();
    })
    config.load();
}
//)

/** help menu **************************************************************/
    function add_help_menu_item() {

        if ($('#latex_envs_help').length > 0) {
            return;
        }
        var menu_item = $('<li/>')
            .insertAfter('#keyboard_shortcuts');
        var menu_link = $('<a/>')
            .text('LaTeX_envs help')
            .attr('title', 'LaTeX_envs documentation')
            .attr('id', "latex_envs_help")
            .attr('href', '/nbextensions/latex_envs/doc/latex_env_doc.html')
            .attr('target', "_blank")
            .appendTo(menu_item);
        $('<i/>')
            .addClass('fa fa-external-link menu-icon pull-right')
            .prependTo(menu_link);
    }


/** LaTeX_envs menu *********************************************************
* Series of sortcuts to environments in latex_envs
****************************************************************************/

function create_latex_menu(callback) {

    if ($('#Latex_envs').length > 0) {
        return;
    }

    $('#help_menu').parent().before('<li id="Latex_envs"/>')
    $('#Latex_envs').addClass('dropdown')
            .append($('<a/>').attr('href', '#')
            .attr('id', 'latex_envs')
            .addClass('dropdown-toogle')
            .attr('data-toggle', "dropdown")
            .attr('aria-expanded', "false")
            .text("LaTeX_envs"))
            .append($('<ul/>')
            .attr('id', 'latex_envs_menu')
            .addClass('dropdown-menu'))

    //for (var i = 0; i < Object.keys(envsLatex).length; i++) {
    for (var p in envsLatex) {
        var current_env_name = envsLatex[p]['name']
        var current_hint = envsLatex[p]['hint']
        var current_env = envsLatex[p]['env']
        var current_shortcut = envsLatex[p]['shortcut']
        var current_id = "env_" + p
        current_env_name = current_shortcut == "" ? current_env_name : current_env_name + '  (' + current_shortcut + ')'

        var menu_item = $('<li/>').appendTo('#latex_envs_menu')
            .attr('id', 'zozo').attr('title', "titre")

        var menu_link = $('<a/>')
            .text(current_env_name)
            .attr('href', '#')
            .attr('title', current_hint)
            .attr('id', current_id)
            .attr('data-text', current_env)
            .attr('onclick', 'insert_text(this);')
            .appendTo(menu_item);

        if (typeof envsLatex[p]['position'] !== "undefined") {
            menu_link.attr('data-position', envsLatex[p]['position'])
        }

        if (current_shortcut !== "") {
            add_edit_shortcuts[current_shortcut] = {
                help: current_hint,
                help_index: 'ht',
                handler: Function('insert_text($(' + '"#' + current_id + '"))')
            }
        }
    }
    callback && callback();
}

/********************************************************************************************
* Definition of a toolbar that enable to select several options:
*		- equations numbered or labelled
* 		- value of initial counter for equations
*		- style of citations call: by number, eg [1,2], by key, eg [perez2001,buss2012], or apalike
* 			eg (Perez et al., 1988)
*		- name of biblio file (if applicable)
*
*********************************************************************************************/
function config_toolbar(callback) {

    if (config_toolbar_present) {
        config_toolbar_present = false;
        $("#LaTeX_envs_toolbar").remove();
        $(site).height($(window).height() - $('#header').height() - $('#footer').height());
        return
    } else {
        config_toolbar_present = true;
    }
    cfg = Jupyter.notebook.metadata.latex_envs

    //local to this function
    var cite_by_icon = { 'number': 'fa-sort-numeric-asc', 'key': 'fa-key', 'apalike': 'fa-pencil-square-o' }
    var cite_by_tmp = cite_by
    var eqLabel_tmp = eqLabelWithNumbers
    var eq_by_icon = { true: 'fa-sort-numeric-asc', false: 'fa-tag' }

    var eqNumtmp = eqNumInitial;

    // Defining the toolbar --------------------------------------------------------------
    var LaTeX_envs_toolbar = $('<div id="LaTeX_envs_toolbar" class="container edit_mode" >')

    var vertical_separator = '&nbsp;&nbsp;<span style="display: inline-block; \
vertical-align:bottom; width: 0; height: 1.8em;border-left:2px solid #cccccc"></span>&nbsp;&nbsp;'

    var biblioLabel = $('<b/>').html('Bibliography&nbsp;')
    var equationLabel = $('<b/>').html('Equations&nbsp;')

    var LaTeX_envs_help_link = $('<a/>').html('LaTeX_envs&nbsp;')
        .css({
            textDecoration: 'none',
            'font-weight': "bold",
            color: 'black'
        })
        .attr('href', '/nbextensions/latex_envs/doc/latex_env_doc.html')
        .attr('target', "_blank")
        .attr('title', 'LaTeX_envs documentation')

    // input bibliography BibTeX filename
    var input_bibFilename = $('<input/>')
        .attr('type', "text")
        .attr('value', bibliofile)
        .attr('id', "biblio")
        .attr('size', '15')
        .attr('title','Enter BibTeX biblio filename (must be present in current directory)')
        .addClass("edit_mode input-xs")
        .css("vertical-align", "middle")

    // input equations numbering offset
    var input_eqNumInitial = $('<input/>')
        .attr('type', "text")
        .attr('value', eqNumtmp)
        .attr('id', "eqnum")
        .attr('title','Equations numbering begins at...')
        .attr('size', '3')
        .addClass("edit_mode")
        .css("vertical-align", "middle")
        .css("text-align", "right")

    // dropdown menu for selecting the citation style
    var bibStyleMenu = $('<div/>').attr('id', 'citeby').addClass('btn-group')
        .attr('title', 'Select references style: numbered, by key, or apa-like')
        .append($('<a/>')
            .addClass("btn btn-default")
            .append($('<i/>')
                .attr('id', "menu-refs").addClass("fa " + cite_by_icon[cite_by_tmp] + " fa-fw"))
            .append('References')
        )
        .append($('<a/>')
            .addClass("btn btn-default dropdown-toggle")
            .attr('data-toggle', "dropdown")
            .attr('href', "#")
            .append($('<span/>').addClass("fa fa-caret-down")))
        .append($('<ul/>').attr('id', 'choice').addClass("dropdown-menu")
            .append($('<li/>')
                .append($('<a/>')
                    .append($('<i/>').addClass("fa fa-sort-numeric-asc fa-fw"))
                    .append('Numbered')
                ))
            .append($('<li/>')
                .append($('<a/>')
                    .append($('<i/>').addClass("fa fa-key fa-fw"))
                    .append('Key')
                )
            )
            .append($('<li/>')
                .append($('<a/>')
                    .append($('<i/>').addClass("fa fa-pencil-square-o fa-fw"))
                    .append('Apa-like')
                )
            )
        )

    // dropdown menu for selecting the numbering style
    var eqLabelStyle = $('<div/>').attr('id', 'eqby').addClass('btn-group')
        .attr('title','Select equations referencing: by number or by label')
        .append($('<a/>')
            .addClass("btn btn-default")
            .append($('<i/>')
                .attr('id', "menu-eqs").addClass("fa " + eq_by_icon[eqLabelWithNumbers] + " fa-fw"))
            .append('Equations')
        )
        .append($('<a/>')
            .addClass("btn btn-default dropdown-toggle")
            .attr('data-toggle', "dropdown")
            .attr('href', "#")
            .append($('<span/>').addClass("fa fa-caret-down")))
        .append($('<ul/>').attr('id', 'choice').addClass("dropdown-menu")
            .append($('<li/>')
                .append($('<a/>')
                    .append($('<i/>').addClass("fa fa-sort-numeric-asc fa-fw"))
                    .append('Numbered')
                ))
            .append($('<li/>')
                .append($('<a/>')
                    .append($('<i/>').addClass("fa fa-tag  fa-fw"))
                    .append('Label')
                )
            )
        )

    // dropdown menu for parameter selection and toggle


    var configMenu = $('<div/>').attr('id', 'cfgby').addClass('btn-group')
        .attr('title', 'Select/Toogle parameters')
        .append($('<a/>')
            .addClass("btn btn-default")
            .append($('<i/>')
                .attr('id', "menu-config").addClass("fa fa-wrench fa-fw"))
            .append('Toggles')
        )
        .append($('<a/>')
            .addClass("btn btn-default dropdown-toggle")
            .attr('data-toggle', "dropdown")
            .attr('href', "#")
            .append($('<span/>').addClass("fa fa-caret-down")))
        .append($('<ul/>').attr('id', 'choice').addClass("dropdown-menu")        
        .attr('min-width', '300px')
            .append($('<li/>') 
                .append($('<a/>')
                    .attr('id','latex_envs_Menu')
                    .text('Show LaTeX menu')
                    .css('width', '200px')
                    .attr('href', '#')
                    .attr('title', 'Toogle visibility of LaTeX_envs menu')
                    .on('click',function (){
                        cfg.LaTeX_envs_menu_present = LaTeX_envs_menu_present = LaTeX_envs_menu_present = !LaTeX_envs_menu_present
                        $('#latex_envs_Menu > .fa').toggleClass('fa-check', LaTeX_envs_menu_present);
                        toggleLatexMenu();
                    })
                    .prepend($('<i/>').addClass('fa menu-icon pull-right'))
                )
            )
            .append($('<li/>')
                .append($('<a/>')
                    .attr('id','labels_anchors_menu')
                    .text('Show Labels anchors')
                    .attr('href', '#')
                    .attr('title', 'Toogle visibility of labels anchors')
                    .on('click',function (){
                        cfg.labels_anchors = labels_anchors = !labels_anchors
                        $('#labels_anchors_menu > .fa').toggleClass('fa-check', labels_anchors);
                        $('.latex_label_anchor').toggle(labels_anchors)
                    })
                    .prepend($('<i/>').addClass('fa menu-icon pull-right'))
                )
            )
             .append($('<li/>')
                .append($('<a/>')
                    .attr('id','latex_user_defs')
                    .text('LaTeX user definitions')
                    .attr('href', '#')
                    .attr('title', 'Load LaTeX user definitions (file latexdefs.tex)')
                    .on('click',function (){
                        cfg.latex_user_defs = latex_user_defs = !latex_user_defs
                        $('#latex_user_defs > .fa').toggleClass('fa-check', latex_user_defs);
                        if (latex_user_defs) {
                            loadLatexUserDefs();
                            setTimeout(function(){ //there is a race condition somewhere
                                init_cells(); 
                                onMarkdownCellRendering();},1000);                            
                        }
                    })
                    .prepend($('<i/>').addClass('fa menu-icon pull-right'))
                )
            )
        )





    // toggle the latex_envs dropdown menu
    var latex_envs_menu_button = $("<a/>")
        .addClass("btn btn-default")
        .attr('href', "#")
        .attr('title', 'Toogle LaTeX_envs menu')
        .css('color', 'black')
        .attr('id', 'toggleLatexMenu')
        .append($("<i/>").addClass('fa fa-caret-square-o-down'))
    
    // close button
    var suicide_button = $("<a/>")
        .addClass("btn btn-default")
        .attr('href', "#")
        .attr('title', 'Close LaTeX_envs toolbar')
        .css('float', 'right')
        .attr('id', 'suicide')
        .attr('title','Close the LaTeX-envs configuration toolbar')
        .append($("<i/>").addClass('fa fa-power-off'))

    // Finally the toolbar itself
    LaTeX_envs_toolbar.append(LaTeX_envs_help_link)
        .append(vertical_separator)
        .append(biblioLabel)
        .append(input_bibFilename)
        .append(bibStyleMenu)
        .append(vertical_separator)
        .append(equationLabel)
        .append(input_eqNumInitial)
        .append(eqLabelStyle)
        .append(vertical_separator)
        .append(latex_envs_menu_button)
        .append(configMenu)
        .append(suicide_button)

    // Appending the new toolbar to the main one
    $('head').append('<style> input:focus {border-color: #66afe9;\
outline: 0; box-shadow: inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px \
rgba(102, 175, 233, 0.6);}</style>')

    $("#maintoolbar-container").append(LaTeX_envs_toolbar);
    $("#LaTeX_envs_toolbar").css({ 'padding': '5px' });

    // Initializing toogles checks
    $('#labels_anchors_menu > .fa').toggleClass('fa-check', labels_anchors)
    $('#latex_envs_Menu > .fa').toggleClass('fa-check', LaTeX_envs_menu_present);
    $('#latex_user_defs > .fa').toggleClass('fa-check', latex_user_defs);
    
    // Now the callback functions --------------------------------------------  

    $('#toggleLatexMenu').on('click', function() {
        cfg.LaTeX_envs_menu_present = LaTeX_envs_menu_present = !LaTeX_envs_menu_present
        toggleLatexMenu();
    })

    $('#citeby').on('click', '.dropdown-menu li a', function() {
        var tmp_text = $(this).text().trim().toLowerCase()
        switch (tmp_text) {
            case 'numbered':
                cite_by_tmp = 'number';
                break;
            case 'key':
                cite_by_tmp = 'key';
                break;
            case 'apa-like':
                cite_by_tmp = 'apalike';
                break;
            default:
                cite_by_tmp = 'apalike';
        }
        $('#menu-refs').removeClass().addClass("fa " + cite_by_icon[cite_by_tmp] + " fa-fw");
        cfg.cite_by = cite_by_tmp //Jupyter.notebook.metadata.latex_envs.cite_by 
        cite_by = cite_by_tmp //Jupyter.notebook.metadata.latex_envs.cite_by 
        init_nb();
    })


    var kmMode = "command";
    $('#biblio').on('focus', function() {
            kmMode = Jupyter.keyboard_manager.mode;
            Jupyter.keyboard_manager.mode = "edit";
        })
        .on('blur', function() { Jupyter.keyboard_manager.mode = kmMode })
        .on('keypress', function(e) {
            if (e.keyCode == 13) {
                $('#biblio').blur();
                cfg.bibliofile = $("#biblio")[0].value;
                bibliofile = $("#biblio")[0].value;
                init_nb();
            }
        })        

    $('#eqby').on('click', '.dropdown-menu li a', function() {
            //console.log($(this).text());
            var tmp_text = $(this).text().trim().toLowerCase()
            switch (tmp_text) {
                case 'numbered':
                    {
                        eqLabel_tmp = true;
                        if (MathJaxDefined) MathJax.Hub.Config({
                            TeX: {
                                equationNumbers: {
                                    autoNumber: "AMS",
                                    useLabelIds: true
                                }
                            }
                        });
                    }
                    break;
                case 'label':
                    {
                        eqLabel_tmp = false;
                        if (MathJaxDefined) MathJax.Hub.Config({
                            TeX: {
                                equationNumbers: {
                                    autoNumber: "none",
                                    useLabelIds: true
                                }
                            }
                        });
                    }
                    break;
                default:
                    {
                        eqLabel_tmp = true;
                        if (MathJaxDefined) MathJax.Hub.Config({
                            TeX: {
                                equationNumbers: {
                                    autoNumber: "AMS",
                                    useLabelIds: true
                                }
                            }
                        });
                    }
            }
            $('#menu-eqs').removeClass().addClass("fa " + eq_by_icon[eqLabel_tmp] + " fa-fw");
            cfg.eqLabelWithNumbers = eqLabel_tmp; //Jupyter.notebook.metadata.latex_envs.eqLabelWithNumbers
            eqLabelWithNumbers = eqLabel_tmp;
            init_cells();
        })


    $('#eqnum').on('focus', function() {
            kmMode = Jupyter.keyboard_manager.mode;
            Jupyter.keyboard_manager.mode = "edit";
        })
        .on('blur', function() { Jupyter.keyboard_manager.mode = kmMode })
        .on('keypress', function(e) {
            if (e.keyCode == 13) {
                $('#eqnum').blur();
                cfg.eqNumInitial = Number($('#eqnum')[0].value); //Jupyter.notebook.metadata.latex_envs.eqNumInitial 
                eqNumInitial = Number($('#eqnum')[0].value);
                init_cells();
            }
        })

    $('#apply').on('click', function() {
            //
            //set the values of global variables
            cite_by = cite_by_tmp //Jupyter.notebook.metadata.latex_envs.cite_by 
            bibliofile = $("#biblio")[0].value //Jupyter.notebook.metadata.latex_envs.bibliofile 
            eqNumInitial = Number($('#eqnum')[0].value)  //Jupyter.notebook.metadata.latex_envs.eqNumInitial
            eqLabelWithNumbers = eqLabel_tmp //Jupyter.notebook.metadata.latex_envs.eqLabelWithNumbers
                //save into notebook's metadata 
            cfg = Jupyter.notebook.metadata.latex_envs
            cfg.cite_by = cite_by_tmp //Jupyter.notebook.metadata.latex_envs.cite_by 
            cfg.bibliofile = $("#biblio")[0].value //Jupyter.notebook.metadata.latex_envs.bibliofile 
            cfg.eqNumInitial = Number($('#eqnum')[0].value) //Jupyter.notebook.metadata.latex_envs.eqNumInitial
            cfg.eqLabelWithNumbers = eqLabel_tmp //Jupyter.notebook.metadata.latex_envs.eqLabelWithNumbers
                //apply all this
            readBibliography(function() {
                init_cells();
                createReferenceSection();
            });

        })
        .tooltip({ title: 'Apply the selected values', trigger: "hover", delay: { show: 500, hide: 50 } });

    $('#suicide').on('click', function() {
        config_toolbar_present = false;
        $("#LaTeX_envs_toolbar").remove();
        $(site).height($(window).height() - $('#header').height() - $('#footer').height())
    })

}
